crawler/urlJobProcessor.js

```diff
@@ -1,36 +1,106 @@
 import { default as fetch } from 'node-fetch';
+import { default as cheerio } from 'cheerio';
+import { default as urlModule } from 'url';
 
 const probes = [];
 
-function configure(/* opts */) {
-  // TODO: Set up probes. They'll be specified somehow in opts and we
-  // should be able to load each probe individually. Maybe something
-  // like `require('./probe-' + probeName + '.js');`
+function configure(opts) {
+  if (!opts.probes || !opts.probes.length) {
+    return Promise.reject(new Error('At least one urlJobProcessor probe must be specified'));
+  }
+
+  const ret = [Promise.resolve()];
+
+  opts.probes.forEach((probeOpts) => {
+    if (!probeOpts.name) {
+      return Promise.reject(new Error('Probe without name encountered'));
+    }
+
+    // We allow the probe to be specified by name or by path. If specified by name,
+    // we expect it to be a known subdir.
+    const path = probeOpts.path || `./urlJobProcessorProbes/${probeOpts.name}.js`;
+    const probeModule = require(path).default;
+
+    if (!probeModule.runProbe) {
+      return Promise.reject(new Error(`Probe ${probeOpts.name} does not implement runProbe`));
+    }
+
+    let p = Promise.resolve();
+    if (probeModule.configure) {
+      p = probeModule.configure(probeOpts.opts || {});
+    }
+
+    ret.push(p.then(() => {
+      probes.push({
+        name: probeOpts.name,
+        probeFn: probeModule.runProbe,
+      });
+    }));
+  });
+
+  return Promise.all(ret);
```

Comment:
If we get concurrency with kue, do we still need it on for the probes?

Web Craftsman | harald@mozilla.com | @digitarald [skype/irc/line] |
+1-650-963-6612 [wapp/telegram]

On Mon, Feb 22, 2016 at 6:20 PM, Tim Abraldes <notifications@github.com>
wrote:

> In crawler/urlJobProcessor.js
> <https://github.com/mozilla/prowac/pull/30#discussion_r53728116>:
>
> > +    }
> > +
> > +    let p = Promise.resolve();
> > +    if (probeModule.configure) {
> > +      p = probeModule.configure(probeOpts.opts || {});
> > +    }
> > +
> > +    ret.push(p.then(() => {
> > +      probes.push({
> > +        name: probeOpts.name,
> > +        probeFn: probeModule.runProbe,
> > +      });
> > +    }));
> > +  });
> > +
> > +  return Promise.all(ret);
>
> Running in parallel will only be a problem if the probes are affecting
> some shared state. So far that's not the case, and it could hurt
> performance to have them run in serial. Maybe we should run them in serial
> only once we have probes that require that? It should be an easy change
> when we decide to make it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mozilla/prowac/pull/30/files#r53728116>.
>